WS2012R2/lisa/stateEngine.ps1

```diff
@@ -817,49 +824,159 @@ function DoSystemDown([System.Xml.XmlElement] $vm, [XML] $xmlData)
+
+    if (-not $xmlData -or $xmlData -isnot [XML])
+    {
+        LogMsg 0 "Error: DoApplyCheckpoint received a null or bad xmlData parameter - VM $($vm.vmName) disabled"
+        $vm.emailSummary += "    DoApplyCheckpoint received a null xmlData parameter - VM disabled<br />"
+        $vm.currentTest = "done"
+        UpdateState $vm $Disabled
+    }
+
+    $testData = GetTestData $vm.currentTest $xmlData
+    if ($testData -is [System.Xml.XmlElement])
+    {
+        # Do not need to recover from checkpoint
+        if (-not $testData.RevertDefaultSnapshot -or $testData.RevertDefaultSnapshot -eq "False")
+        {
+            LogMsg 9 "Info : noCheckpoint is not configured or set to True."
+            if (-not (VerifyTestResourcesExist $vm $testData))
             {
-                if (-not (VerifyTestResourcesExist $vm $testData))
-                {
-                    #
-                    # One or more resources used by the VM or test case does not exist - fail the test
-                    #
-                    $testName = $testData.testName
-                    $vm.emailSummary += ("    Test {0, -25} : {1}<br />" -f ${testName}, "Failed")
-                    $vm.emailSummary += "          Missing resources<br />"
-                    return
-                }
+                #
+                # One or more resources used by the VM or test case does not exist - fail the test
+                #
+                $testName = $testData.testName
+                $vm.emailSummary += ("    Test {0, -25} : {1}<br />" -f ${testName}, "Failed")
+                $vm.emailSummary += "          Missing resources<br />"
+                $vm.currentTest = "done"
+                UpdateState $vm $Disabled
+            }
+        }
+        # Case requires a fresh new state VM to run.
+        else
+        {
+            #
+            # Reset the VM to a snapshot to put the VM in a known state.  The default name is
+            # ICABase.  This can be overridden by the global.defaultSnapshot in the global section
+            # and then by the vmSnapshotName in the VM definition.
+            #
+            $snapshotName = "ICABase"
 
-                if ($vm.preStartConfig -or $testData.setupScript)
+            if ($xmlData.config.global.defaultSnapshot)
+            {
+                $snapshotName = $xmlData.config.global.defaultSnapshot
+                LogMsg 5 "Info : $($vm.vmName) Over-riding default snapshotName from global section to $snapshotName"
+            }
+
+            if ($vm.vmSnapshotName)
+            {
+                $snapshotName = $vm.vmSnapshotName
+                LogMsg 5 "Info : $($vm.vmName) Over-riding default snapshotName from VM section to $snapshotName"
+            }
+
+            #
+            # Find the snapshot we need and apply the snapshot
+            #
+            $snapshotFound = $false
+            $snaps = Get-VMSnapshot $vm.vmName -ComputerName $vm.hvServer
```

Comment:
You should use "Get-VMSnapshot -Name $snapshotName  -VMName $vm.vmName". This will avoid the need for the foreach loop in the next part.